### ISSUE TITLE
C FFI Compilation bug

### ISSUE THREAD

**Issue by dfordivam** [WebGHC/ghc#6]
For this code (from bytestring) clang gives error

```
foreign import ccall unsafe "static stdlib.h &free" c_free_finalizer
    :: FunPtr (Ptr Word8 -> IO ())


./ghc25458_0/ghc_4.hc:153:42: error: use of undeclared identifier 'free'
(W_)&base_GHCziPtr_FunPtr_con_info, (W_)&free
```

It is solved by adding the appropriate includes in ghc_4.hc

     #include <stdlib.h>

**Comment by ElvishJerricco** [comment]
Yes, the C FFI in Haskell supports using `&` to indicate you want a reference to the symbol instead of calling it. This does *not* work in `ccall` (EDIT: on wasm, that is), and AFAICT, there's no general way to make it work on wasm due to the strict type requirements. This *does* work with the `capi` FFI though. In fact, I believe the `capi` FFI should be preferred virtually always on wasm.

**Comment by ElvishJerricco** [comment]
See e.g.: https://github.com/WebGHC/ghc/commit/7f1605d3ce7e5db14b851344f1feabf731399324#diff-79ec7cacd18d11f92a5069572e68989d

**Comment by dfordivam** [comment]
Ok.. this ought to be documented somewhere..
But for the time being lets keep this issue open..
